astra/src/main/java/com/slack/astra/chunk/ReadOnlyChunkImpl.java

```
@@ -228,6 +231,48 @@ public CacheNodeAssignment getCacheNodeAssignment() {
     return assignment;
   }
 
+  private boolean validateS3vsLocalDownLoad() {
+    // check if the number of files in S3 matches the local directory
+    Map<String, Long> filesWithSizeInS3 = blobStore.listFilesWithSize(snapshotMetadata.snapshotId);
+
+    Map<String, Long> localFiles;
+    try (Stream<Path> fileList = Files.list(dataDirectory)) {
+      localFiles =
+          fileList
+              .filter(Files::isRegularFile)
+              .collect(
+                  Collectors.toMap(
+                      path ->
+                          dataDirectory.relativize(path).toString().replace(File.separator, "/"),
+                      path -> path.toFile().length()));
+    } catch (IOException e) {
+      throw new RuntimeException(
+          String.format("Error reading local files in directory %s", dataDirectory), e);
+    }
+    if (localFiles.size() != filesWithSizeInS3.size()) {
+      LOG.error(
+          String.format(
+              "Mismatch in number of files in S3 (%s) and local directory (%s) for snapshot %s",
+              filesWithSizeInS3.size(), localFiles.size(), snapshotMetadata.toString()));
+      return false;
+    }
+
+    for (Map.Entry<String, Long> entry : filesWithSizeInS3.entrySet()) {
+      String s3Path = entry.getKey();
+      long s3Size = entry.getValue();
+      String fileName = Paths.get(s3Path).getFileName().toString();
+
+      if (!localFiles.containsKey(fileName) || !localFiles.get(fileName).equals(s3Size)) {
+        LOG.error(
+            String.format(
+                "Mismatch for file %s in S3 and local directory of size %s for snapshot %s",
+                s3Path, s3Size, snapshotMetadata.toString()));
+        return false;
```

It would be nice if instead of returning on the first failure it logged a list of the files with differing sizes.